### PR TITLE
Exclude some fp16 support files from ort minimal build

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
### Description

As title.

Note: also found out some other diffs that can be potentially improved for binary size. e.g. transpose_helper.cc - pass in a set of element data type lists in use (we don't have that available yet) instead of the default full list, etc. did the obvious change for now.

### Motivation and Context

Reduce minimal build binary size for mobile to meet office team requirement.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc [hunk 1]
```diff
@@ -159,6 +159,7 @@ Status RegisterNchwcKernels(KernelRegistry& kernel_registry) {
   return Status::OK();
 }
 
+#if !defined(ORT_MINIMAL_BUILD)
```
Comment (edgchen1) [inline onnxruntime/contrib_ops/cpu/cpu_contrib_kernels.cc:162]: should this be `#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)` to be consistent with the kernel files?

File: onnxruntime/core/mlas/lib/halfgemm.h [hunk 2]
```diff
@@ -513,3 +515,5 @@ MlasHalfGemmGetDispatch()
     return &MlasHalfGemmDispatchDefault;
 #endif
 }
+
+#endif
```
Comment (edgchen1) [inline onnxruntime/core/mlas/lib/halfgemm.h:519]: general: it's helpful to have a comment with the condition on the `#endif` especially when the `#if` is further away

```suggestion
#endif  // !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
```

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 3]
```diff
@@ -2395,6 +2395,7 @@ Status RegisterOnnxOperatorKernels(KernelRegistry& kernel_registry) {
   return Status::OK();
 }
 
+#if !defined(ORT_MINIMAL_BUILD)
```
Comment (edgchen1) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc:2398]: should this be `#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)`?

File: onnxruntime/core/mlas/lib/halfgemm.h [hunk 4]
```diff
@@ -30,6 +30,8 @@ Module Name:
         MLAS_HALF_GEMM_STRIDES Strides{128, 128, 128};
 --*/
 
+#if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/mlas/lib/halfgemm.h]: there are a few other places where the `#ifdef` is before `#pragma once`. e.g., test_fp16.h and test_halfgemm.h

**Review by edgchen1 — APPROVED** [review]

**Comment by YUNQIUGUO** [comment]
close this pr. substitute with #17259

**Pull request closed without merging**